File: php/monsterhunterworld_sdk.php
<?php
declare(strict_types=1);

// MonsterHunterWorld SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class MonsterHunterWorldSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new MonsterHunterWorldUtility();
        $this->_utility = $utility;

        $config = MonsterHunterWorldConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = MonsterHunterWorldHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = MonsterHunterWorldHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, MonsterHunterWorldFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return MonsterHunterWorldUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = MonsterHunterWorldHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = MonsterHunterWorldHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = MonsterHunterWorldHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new MonsterHunterWorldSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = MonsterHunterWorldHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = MonsterHunterWorldHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_ailment = null;

    // Canonical facade: $client->Ailment()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->ailment()
    // resolves here too.
    public function Ailment($data = null)
    {
        require_once __DIR__ . '/entity/ailment_entity.php';
        if ($data === null) {
            if ($this->_ailment === null) {
                $this->_ailment = new AilmentEntity($this, null);
            }
            return $this->_ailment;
        }
        return new AilmentEntity($this, $data);
    }


    private $_armor = null;

    // Canonical facade: $client->Armor()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->armor()
    // resolves here too.
    public function Armor($data = null)
    {
        require_once __DIR__ . '/entity/armor_entity.php';
        if ($data === null) {
            if ($this->_armor === null) {
                $this->_armor = new ArmorEntity($this, null);
            }
            return $this->_armor;
        }
        return new ArmorEntity($this, $data);
    }


    private $_armor_set = null;

    // Canonical facade: $client->ArmorSet()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->armor_set()
    // resolves here too.
    public function ArmorSet($data = null)
    {
        require_once __DIR__ . '/entity/armor_set_entity.php';
        if ($data === null) {
            if ($this->_armor_set === null) {
                $this->_armor_set = new ArmorSetEntity($this, null);
            }
            return $this->_armor_set;
        }
        return new ArmorSetEntity($this, $data);
    }


    private $_charm = null;

    // Canonical facade: $client->Charm()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->charm()
    // resolves here too.
    public function Charm($data = null)
    {
        require_once __DIR__ . '/entity/charm_entity.php';
        if ($data === null) {
            if ($this->_charm === null) {
                $this->_charm = new CharmEntity($this, null);
            }
            return $this->_charm;
        }
        return new CharmEntity($this, $data);
    }


    private $_decoration = null;

    // Canonical facade: $client->Decoration()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->decoration()
    // resolves here too.
    public function Decoration($data = null)
    {
        require_once __DIR__ . '/entity/decoration_entity.php';
        if ($data === null) {
            if ($this->_decoration === null) {
                $this->_decoration = new DecorationEntity($this, null);
            }
            return $this->_decoration;
        }
        return new DecorationEntity($this, $data);
    }


    private $_event = null;

    // Canonical facade: $client->Event()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->event()
    // resolves here too.
    public function Event($data = null)
    {
        require_once __DIR__ . '/entity/event_entity.php';
        if ($data === null) {
            if ($this->_event === null) {
                $this->_event = new EventEntity($this, null);
            }
            return $this->_event;
        }
        return new EventEntity($this, $data);
    }


    private $_item = null;

    // Canonical facade: $client->Item()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->item()
    // resolves here too.
    public function Item($data = null)
    {
        require_once __DIR__ . '/entity/item_entity.php';
        if ($data === null) {
            if ($this->_item === null) {
                $this->_item = new ItemEntity($this, null);
            }
            return $this->_item;
        }
        return new ItemEntity($this, $data);
    }


    private $_location = null;

    // Canonical facade: $client->Location()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->location()
    // resolves here too.
    public function Location($data = null)
    {
        require_once __DIR__ . '/entity/location_entity.php';
        if ($data === null) {
            if ($this->_location === null) {
                $this->_location = new LocationEntity($this, null);
            }
            return $this->_location;
        }
        return new LocationEntity($this, $data);
    }


    private $_monster = null;

    // Canonical facade: $client->Monster()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->monster()
    // resolves here too.
    public function Monster($data = null)
    {
        require_once __DIR__ . '/entity/monster_entity.php';
        if ($data === null) {
            if ($this->_monster === null) {
                $this->_monster = new MonsterEntity($this, null);
            }
            return $this->_monster;
        }
        return new MonsterEntity($this, $data);
    }


    private $_motion_value = null;

    // Canonical facade: $client->MotionValue()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->motion_value()
    // resolves here too.
    public function MotionValue($data = null)
    {
        require_once __DIR__ . '/entity/motion_value_entity.php';
        if ($data === null) {
            if ($this->_motion_value === null) {
                $this->_motion_value = new MotionValueEntity($this, null);
            }
            return $this->_motion_value;
        }
        return new MotionValueEntity($this, $data);
    }


    private $_skill = null;

    // Canonical facade: $client->Skill()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->skill()
    // resolves here too.
    public function Skill($data = null)
    {
        require_once __DIR__ . '/entity/skill_entity.php';
        if ($data === null) {
            if ($this->_skill === null) {
                $this->_skill = new SkillEntity($this, null);
            }
            return $this->_skill;
        }
        return new SkillEntity($this, $data);
    }


    private $_weapon = null;

    // Canonical facade: $client->Weapon()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->weapon()
    // resolves here too.
    public function Weapon($data = null)
    {
        require_once __DIR__ . '/entity/weapon_entity.php';
        if ($data === null) {
            if ($this->_weapon === null) {
                $this->_weapon = new WeaponEntity($this, null);
            }
            return $this->_weapon;
        }
        return new WeaponEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new MonsterHunterWorldSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}

File: lua/monster-hunter-world_sdk.lua
-- MonsterHunterWorld SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local MonsterHunterWorldSDK = {}
MonsterHunterWorldSDK.__index = MonsterHunterWorldSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

MonsterHunterWorldSDK._make_feature = _make_feature


function MonsterHunterWorldSDK.new(options)
  local self = setmetatable({}, MonsterHunterWorldSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function MonsterHunterWorldSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function MonsterHunterWorldSDK:get_utility()
  return Utility.copy(self._utility)
end


function MonsterHunterWorldSDK:get_root_ctx()
  return self._rootctx
end


function MonsterHunterWorldSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function MonsterHunterWorldSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:Ailment():list() / client:Ailment():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Ailment(data)
  local EntityMod = require("entity.ailment_entity")
  if data == nil then
    if self._ailment == nil then
      self._ailment = EntityMod.new(self, nil)
    end
    return self._ailment
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Armor():list() / client:Armor():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Armor(data)
  local EntityMod = require("entity.armor_entity")
  if data == nil then
    if self._armor == nil then
      self._armor = EntityMod.new(self, nil)
    end
    return self._armor
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ArmorSet():list() / client:ArmorSet():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:ArmorSet(data)
  local EntityMod = require("entity.armor_set_entity")
  if data == nil then
    if self._armor_set == nil then
      self._armor_set = EntityMod.new(self, nil)
    end
    return self._armor_set
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Charm():list() / client:Charm():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Charm(data)
  local EntityMod = require("entity.charm_entity")
  if data == nil then
    if self._charm == nil then
      self._charm = EntityMod.new(self, nil)
    end
    return self._charm
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Decoration():list() / client:Decoration():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Decoration(data)
  local EntityMod = require("entity.decoration_entity")
  if data == nil then
    if self._decoration == nil then
      self._decoration = EntityMod.new(self, nil)
    end
    return self._decoration
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Event():list() / client:Event():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Event(data)
  local EntityMod = require("entity.event_entity")
  if data == nil then
    if self._event == nil then
      self._event = EntityMod.new(self, nil)
    end
    return self._event
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Item():list() / client:Item():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Item(data)
  local EntityMod = require("entity.item_entity")
  if data == nil then
    if self._item == nil then
      self._item = EntityMod.new(self, nil)
    end
    return self._item
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Location():list() / client:Location():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Location(data)
  local EntityMod = require("entity.location_entity")
  if data == nil then
    if self._location == nil then
      self._location = EntityMod.new(self, nil)
    end
    return self._location
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Monster():list() / client:Monster():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Monster(data)
  local EntityMod = require("entity.monster_entity")
  if data == nil then
    if self._monster == nil then
      self._monster = EntityMod.new(self, nil)
    end
    return self._monster
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:MotionValue():list() / client:MotionValue():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:MotionValue(data)
  local EntityMod = require("entity.motion_value_entity")
  if data == nil then
    if self._motion_value == nil then
      self._motion_value = EntityMod.new(self, nil)
    end
    return self._motion_value
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Skill():list() / client:Skill():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Skill(data)
  local EntityMod = require("entity.skill_entity")
  if data == nil then
    if self._skill == nil then
      self._skill = EntityMod.new(self, nil)
    end
    return self._skill
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Weapon():list() / client:Weapon():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function MonsterHunterWorldSDK:Weapon(data)
  local EntityMod = require("entity.weapon_entity")
  if data == nil then
    if self._weapon == nil then
      self._weapon = EntityMod.new(self, nil)
    end
    return self._weapon
  end
  return EntityMod.new(self, data)
end




function MonsterHunterWorldSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = MonsterHunterWorldSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return MonsterHunterWorldSDK
